compiler/src/dotty/tools/dotc/transform/ShortcutImplicits.scala

```diff
@@ -0,0 +1,165 @@
+ *
+ *  is expanded to:
+ *
+ *     def m(xs: Ts): IF
+ *     def m$direct(xs: Ts)(ys: Us): R
+ *
+ *  (2) A reference `qual.apply` where `qual` has implicit function type and
+ *  `qual` refers to a method `m` is rewritten to a reference to `m$direct`,
+ *  keeping the same type and value arguments as they are found in `qual`.
+ */
+class ShortcutImplicits extends MiniPhase with IdentityDenotTransformer { thisTransform =>
+  import tpd._
+
+  override def phaseName: String = "shortcutImplicits"
+  val treeTransform = new Transform
+
+  /** If this option true, we don't specialize symbols that are known to be only
+   *  targets of monomorphic calls.
+   *  The reason for this option is that benchmarks show that on the JVM for monomorphic dispatch
+   *  scenarios inlining and escape analysis can often remove all calling overhead, so we might as
+   *  well not duplicate the code. We need more experience to decide on the best setting of this option.
+   */
+  final val specializeMonoTargets = true
+
+  class Transform extends TreeTransform {
+    def phase = thisTransform
+
+    override def prepareForUnit(tree: Tree)(implicit ctx: Context) = new Transform
+
+    /** A map to cache mapping local methods to their direct counterparts.
+     *  A fresh map is created for each unit.
+     */
+    private val directMeth = new mutable.HashMap[Symbol, Symbol]
+
+    /** Should `sym` get a ..$direct companion?
+     *  This is the case if (1) `sym` is a method with an implicit function type as final result type.
+     *  However if `specializeMonoTargets` is true, we exclude symbols that are known
+     *  to be only targets of monomorphic calls because they are effectively
+     *  final and don't override anything.
+     */
+    private def shouldBeSpecialized(sym: Symbol)(implicit ctx: Context) =
+      sym.is(Method, butNot = Accessor) &&
+      defn.isImplicitFunctionType(sym.info.finalResultType) &&
+      (specializeMonoTargets || !sym.isEffectivelyFinal || sym.allOverriddenSymbols.nonEmpty)
+
+    /** @pre    The type's final result type is an implicit function type `implicit Ts => R`.
+     *  @return The type of the `apply` member of `implicit Ts => R`.
+     */
+    private def directInfo(info: Type)(implicit ctx: Context): Type = info match {
+      case info: PolyType   => info.derivedPolyType(resType = directInfo(info.resultType))
+      case info: MethodType => info.derivedMethodType(resType = directInfo(info.resultType))
+      case info: ExprType   => directInfo(info.resultType)
+      case info             => info.member(nme.apply).info
+    }
+
+    /** A new `m$direct` method to accompany the given method `m` */
+    private def newDirectMethod(sym: Symbol)(implicit ctx: Context): Symbol = {
+      val direct = sym.copy(
+        name = sym.name.directName,
+        flags = sym.flags | Synthetic,
+        info = directInfo(sym.info))
+      if (direct.allOverriddenSymbols.isEmpty) direct.resetFlag(Override)
+      direct
```

Comment:
It will invoke infoTransforms, but this phase doesn't register one, it forcefully updates denotations.